src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioDiagnosticListSuppressionStateService.cs

```diff
@@ -0,0 +1,253 @@
+    /// </summary>
+    [Export(typeof(IVisualStudioDiagnosticListSuppressionStateService))]
+    internal class VisualStudioDiagnosticListSuppressionStateService : IVisualStudioDiagnosticListSuppressionStateService
+    {
+        private readonly VisualStudioWorkspace _workspace;
+        private readonly IVsUIShell _shellService;
+        private readonly IWpfTableControl _tableControl;
+
+        private int _selectedActiveItems;
+        private int _selectedSuppressedItems;
+        private int _selectedRoslynItems;
+        private int _selectedCompilerDiagnosticItems;
+        private int _selectedNonSuppressionStateItems;
+
+        [ImportingConstructor]
+        public VisualStudioDiagnosticListSuppressionStateService(
+            SVsServiceProvider serviceProvider,
+            VisualStudioWorkspace workspace)
+        {
+            _workspace = workspace;
+            _shellService = (IVsUIShell)serviceProvider.GetService(typeof(SVsUIShell));
+            var errorList = serviceProvider.GetService(typeof(SVsErrorList)) as IErrorList;
+            _tableControl = errorList?.TableControl;
+
+            ClearState();
+            InitializeFromTableControlIfNeeded();
+        }
+
+        private int SelectedItems => _selectedActiveItems + _selectedSuppressedItems + _selectedNonSuppressionStateItems;
+
+        // If we can suppress either in source or in suppression file, we enable suppress context menu.
+        public bool CanSuppressSelectedEntries => CanSuppressSelectedEntriesInSource || CanSuppressSelectedEntriesInSuppressionFiles;
+
+        // If only suppressed items are selected, we enable remove suppressions.
+        public bool CanRemoveSuppressionsSelectedEntries => _selectedActiveItems == 0 && _selectedSuppressedItems > 0;
+
+        // If only Roslyn active items are selected, we enable suppress in source.
+        public bool CanSuppressSelectedEntriesInSource => _selectedActiveItems > 0 &&
+            _selectedSuppressedItems == 0 &&
+            _selectedRoslynItems == _selectedActiveItems;
+
+        // If only active items are selected, and there is at least one Roslyn item, we enable suppress in suppression file.
+        // Also, compiler diagnostics cannot be suppressed in suppression file, so there must be at least one non-compiler item.
+        public bool CanSuppressSelectedEntriesInSuppressionFiles => _selectedActiveItems > 0 &&
+            _selectedSuppressedItems == 0 &&
+            (_selectedRoslynItems - _selectedCompilerDiagnosticItems) > 0;
+
+        private void ClearState()
+        {
+            _selectedActiveItems = 0;
+            _selectedSuppressedItems = 0;
+            _selectedRoslynItems = 0;
+            _selectedCompilerDiagnosticItems = 0;
+            _selectedNonSuppressionStateItems = 0;
+        }
+
+        private void InitializeFromTableControlIfNeeded()
+        {
+            if (_tableControl == null)
```

Comment:
yes, basically when querying status of the suppression commands. most callbacks just look at newly selected or unselected entries, so it should be fast. The check at the start of this method is in fact true almost all the time, so we don't traverse all entries.